### ISSUE TITLE
[Advise] This package should not be used for big production apps

### ISSUE THREAD

**Issue by nombrekeff** [diamondio/better-queue#52]
Hi, I'm sorry for filing so many issues lately but I was using it for a work project and needed some feedback.

> First of all I really apreciate the work put into this project, and this is just a constructive advise.

### Let me explain
Here is the thing, I think it's a really **great package** that serves a purpose, but I think it's **not stable** enough to be considered stable nor be used in production or in big apps. At least not for now, mostly because of the **stores**, its not 100% reliable, and it fails more than it works... Although memory store works pretty good _(I can explain this in detail if this issue goes through)_.

> In the end I switched to using **RabbitMQ** as it's more solid and stable for bigger and profesional applications.

I'm willing to **contribute** to this package as I really think it has its purpose, and could be very powerfull... we could also get inspiration from RabbitMQ...

Let me know what you think, and if this has already been talked about or I'm giving it more importance than it has. Also If you'd like me to collaborate in some way.

👋 Cheers!

**Comment by benfavre** [comment]
Let's rephrase this.
This package is great !
Carefull tho ... There are many more robust solutions if you are looking for a persistant (distributed or not) queue with a store.

**Comment by nombrekeff** [comment]
Yup, totally agree with you, it's a really great package for personal or small scale projects, for bigger projects, other robust solutions are available...

**Comment by benfavre** [comment]
What does this even mean ?

We use it for CLI jobs that need to process many lines of huge csv files asynchronously.
If something fails we either throw it await or the process needs to be requeued anyway, this is in "production".

This is how most things work in modern architectures.
The CLI jobs themselves are managed by a robust task queue, yes.

The "size" or "commercial" aspect has absolutely nothing to do with the viability of a package.
Noobs will read this and think there is a inherent issue with using something like this.

**Comment by nombrekeff** [comment]
> We use it for CLI jobs that need to process many lines of huge csv files asynchronously.
If something fails we either throw it await or the process needs to be requeued anyway, this is in "production".

Where does this come from? Or who post it it? 

> The "size" or "commercial" aspect has absolutely nothing to do with the viability of a package.

Agree, In no way I'm deprecating this package or anything similar. I really appreciate what has been done here!!
I'm sorry if I gave that impresion, it was not my intention... 😢 

> Noobs will read this and think there is a inherent issue with using something like this.  

Totally agree, my intention was not to confuse Noobs but to advise them that for bigger projects I would not recommend using this package. But for CLI applications and small projects it's totally viable and quite useful actually. 

I'm sorry if I've caused any issue or misguided anyone to not use this package, it was not my intention...

**Comment by benfavre** [comment]
Yes, I meant that we use it.

I think it's important to not give people the impression something is bad when it's more that it's not suited for everything.

I know that me 10 years ago reading your comment would look for another package, when in reality your comment is about architecture.

Don't be sorry, you have engaged in a conversation that might be usefull to other people.

Have a nice day.

**Comment by nombrekeff** [comment]
Yes, that's true, I will try to make the issue more clear for newcomers.

I think I would've thought the same thing a couple of years ago 😄

It would be nice to hear the opinion of a maintainer, to see what they think.

Cheers man! Have a nice day too.

**Comment by leanderlee** [comment]
Yes better-queue won't have the durability of rabbitmq, for example. But it's still useful for a lot of cases, even in production. The code is now a few years old and can do with a refresh, but I don't really have time to modernize it.

Perhaps others in the community would be interested in working on a more performant version of this library that would be able to offer higher durability guarantees.

**Comment by leanderlee** [comment]
Is there a concrete example of the shortcomings of better-queue? This might be helpful for people to people looking to make a decision.

**Comment by nombrekeff** [comment]
Yup, thanks, that's definitely a better title  👍 


> ... But it's still useful for a lot of cases, even in production.

I totally agree, I should of said Big Scale Applications or for expandable/distributed systems where performance and reliability are essential, instead... I will explain in better detail and rephrase the issue description accordingly when I find a little bit of time...

> Is there a concrete example of the shortcomings of better-queue?

I will try explaining what my shortcomings where, what limitations I had and why I chose to switch to rabbitmq.

> ...but I don't really have time to modernize it.
I might have a look at the current source code and have a try on it.

Are there more/any active maintainers at  the moment?

**Comment by LinusU** [comment]
> Are there more/any active maintainers at the moment?

I read all comments and try to respond to pull requests asap ☺️

**Comment by nombrekeff** [comment]
Cool, I may ask some questions about the project, when I start testing/fiddling the package

**Comment by turbopasi** [comment]
Reading through the comments now I am a bit confused whether this is package is safe to use or not. The only thing I get from here is it's not stable enough and it's not 100 % reliable . Does that mean it skips tasks randomly ? I am using it in a middle-sized project on a video-encoding-server ( video gets uploaded and stored, then encoded with ffmpeg ) . 

Are there any examples, where better-queue might fail ?

**Comment by nombrekeff** [comment]
Hey, it is stable in my experience.

My point was just noting that for big projects it might not be the best option. But it does the job very well, just not as scalable as let's say RabbitMQ or similar. 

If you intend to scale up or become decentralized this might not work for you in the long term, but for small-mid scale apps, I think it will do the job just fine.

> Are there any examples, where better-queue might fail?

For me, it became a problem when scaling up and migrating to microservices. Because we needed to have 2 instances of the queuing system that talked to each other and were synchronized, as to have a fallback in case one of them stopped or something.
We also needed the queue to be accessible from various parts of the system, so we could not do it with this library.

**Comment by nombrekeff** [comment]
I've updated a bit the issue's body, clearing some misleading info. Hopefully, it will be easier for people to understand the issue.

**Comment by turbopasi** [comment]
I was trying to implement my own storage (local store with lowdb) and I am starting to have troubles as well now. Queue "randomly" stops processing although the queue is full with tasks. To implement a custom storage should probably be documented better. I guess when using the default memory storage all works fine.

**Comment by apiel** [comment]
You can't really compare such library to RabbitMQ.

RabbitMQ is very complicated to setup and you need to understand all the possibles configurations and protocols available to fit your need. Setting up such infrastructure require time and practice, really not for beginner.

The kind of queue like better-queue, should be very simple to understand, to setup and use, and should be intended for simple use-case where you want to decouple your code / task.
